tree-building/tree_building.go

```diff
@@ -0,0 +1,128 @@
+package tree
+
+import "fmt"
+import "sort"
+
+type Record struct {
+	ID     int
+	Parent int
+}
+
+type Node struct {
+	ID       int
+	Children []*Node
+}
+
+var continuos bool
+
+func Build(records []Record) (*Node, error) {
+	records = orderRecords(records)
+	if isDisContinuous(records) {
+		return nil, fmt.Errorf("Not continuous")
+	}
+	result, error := BuildTree(records, true)
+	return result, error
+}
+
+func BuildTree(records []Record, ordered bool) (*Node, error) {
+	//Wrote this as a separate function to avoid isDisContinuous check for subtrees
+	if !ordered {
+		records = orderRecords(records)
+	}
+	if len(records) == 0 {
+		return nil, nil
+	}
+	var parentPresent bool //To check if multiple parents are present
+	var result Node
+	tree := subTree(records) // To get subtrees of the element
+	for _, v := range records {
+		err := handleErrors(records, v, parentPresent)
+		if err != "" {
+			return nil, fmt.Errorf(err)
+		}
+		if v.ID == v.Parent {
+			result.ID = v.ID
+			parentPresent = true
+		} else {
+			if result.ID == v.Parent {
+				sub_child, _ := BuildTree(append(tree[v.ID], Record{v.ID, v.ID}), false)
+				result.Children = append(result.Children, sub_child)
+			}
+		}
+	}
+	if !parentPresent {
+		return nil, fmt.Errorf("No Parent")
+	}
+	orderChildren(result.Children)
+	return &result, nil
+}
+
+func subTree(records []Record) map[int][]Record {
+	//Function to get all children for a node
+	tree := make(map[int][]Record)
+	for _, v := range records {
+		if v.ID != v.Parent {
+			if len(tree[v.Parent]) == 0 {
+				tree[v.Parent] = []Record{v}
+			} else {
+				tree[v.Parent] = append(tree[v.Parent], v)
+			}
+
+		}
+	}
+	return tree
+
+}
+
+func contains(s []Record, e Record) bool {
+	// Function to detect duplicates
+	var counter int
+	for _, a := range s {
+		if a == e {
+			counter = counter + 1
```

Comment:
```suggestion
			counter++
```
Also, on the next line, you could just do
```go
if counter == 2 {
    return true
}
```
and `return false` at the end. This would short circuit the loop as soon as a dupe is found.